File: web/src/types/order.ts
import { z } from 'zod'

export const orderItemInSchema = z.object({
  product_id: z.string(),
  description: z.string().optional(),
  quantity: z.number().gt(0),
  unit_price: z.number().min(0),
  line_discount_rate: z.number().min(0).max(100).optional(),
  tax_rate: z.number().min(0).max(100).optional(),
})

export type OrderItemIn = z.infer<typeof orderItemInSchema>

export const orderSchema = z.object({
  id: z.string(),
  number: z.string(),
  partner_id: z.string(),
  currency: z.string(),
  status: z.enum(['NEW', 'CONFIRMED', 'FULFILLED', 'CANCELLED']),
  order_date: z.string(),
  notes: z.string().optional(),
  discount_rate: z.number().min(0).max(100),
  subtotal: z.number(),
  tax_total: z.number(),
  grand_total: z.number(),
  created_at_utc: z.string(),
})

export type Order = z.infer<typeof orderSchema>

export const orderDetailSchema = orderSchema.extend({
  items: z
    .array(
      orderItemInSchema.extend({
        line_subtotal: z.number(),
        line_tax: z.number(),
        line_total: z.number(),
      })
    )
    .min(1),
})

export type OrderDetail = z.infer<typeof orderDetailSchema>

export const orderCreateSchema = z.object({
  partner_id: z.string(),
  currency: z.string(),
  order_date: z.string(),
  notes: z.string().optional(),
  discount_rate: z.number().min(0).max(100).default(0),
  items: z.array(orderItemInSchema).min(1),
})

export type OrderCreate = z.infer<typeof orderCreateSchema>

export const orderUpdateSchema = orderCreateSchema
export type OrderUpdate = z.infer<typeof orderUpdateSchema>

File: backend/app/schemas/order.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import PageMeta


class OrderItemBase(BaseModel):
    product_id: UUID
    description: str | None = None
    quantity: Decimal = Field(..., gt=0)
    unit_price: Decimal = Field(..., ge=0)
    line_discount_rate: Decimal = Field(0, ge=0, le=100)
    tax_rate: Decimal = Field(20, ge=0, le=100)


class OrderItemCreate(OrderItemBase):
    pass


class OrderBase(BaseModel):
    partner_id: UUID
    project_name: str | None = None
    delivery_date: date | None = None
    status: str | None = "TEKLIF"
    discount_rate: Decimal = Field(0, ge=0, le=100)
    notes: str | None = None
    items: list[OrderItemCreate]


class OrderCreate(OrderBase):
    pass


class OrderUpdate(BaseModel):
    project_name: str | None = None
    delivery_date: date | None = None
    status: str | None = None
    discount_rate: Decimal | None = Field(None, ge=0, le=100)
    notes: str | None = None
    items: list[OrderItemCreate] | None = None


class OrderItemPublic(BaseModel):
    id: UUID
    product_id: UUID
    description: str | None
    quantity: Decimal
    unit_price: Decimal
    line_discount_rate: Decimal
    tax_rate: Decimal
    line_subtotal: Decimal
    line_tax: Decimal
    line_total: Decimal

    model_config = ConfigDict(from_attributes=True)


class OrderPublic(BaseModel):
    id: UUID
    organization_id: UUID
    number: str
    partner_id: UUID
    project_name: str | None
    delivery_date: date | None
    status: str
    discount_rate: Decimal
    subtotal: Decimal
    tax_total: Decimal
    grand_total: Decimal
    notes: str | None
    created_at_utc: datetime
    items: list[OrderItemPublic]

    model_config = ConfigDict(from_attributes=True)


class OrderListResponse(PageMeta):
    items: list[OrderPublic]

